### PR TITLE
support interface for non-block requests

### LINKED ISSUE #45: Support non-block requests
the first top-level interface of caboose is as a blockstore.

we should work with @aschmahmann  to understand what requests for car files rather than blocks might look like as they pass through this library.

My first inclination is something like

```golang
Fetch(urlPath string, cb func(io.ReadCloser) error) error
```

where `cb` would validate if a response matched expected response (and would be called 0+ times, and at most 1 time where it returned a `nil` error)

### PR DESCRIPTION
**Pull request by willscott**
`fetchResource` is an underlying fetch request that can be called in a block or non-block context.

This isn't a complete abstraction, as the `fetchWith` logic will need to be also be applied on the in-development car interface.

This currently separates an individual block request / validation from the the internal request and tracking.

fix #45

### CODE REVIEW

File: metrics.go
Comment (lidel) [inline metrics.go:81]: We need `fetch_duration_car_peer_success` version as well (to match what we already have for blocks).
Even if we have no retries today, we should add these metrics from day one, so we can add them to Grafana.

File: metrics.go
Comment (lidel) [inline metrics.go:93]: We need `fetch_duration_car_peer_failure` version as well (to match what we already have for blocks)

**Review by lidel — CHANGES_REQUESTED** [review]

File: caboose.go
Comment (aarshkshah1992) [reply caboose.go:77]: Looks like this error type isn't actually used anywhere. Is it the caller's responsibility to return this error from the given callback to `Fetch` so we can do stuff accordingly ?

File: caboose.go
Comment (aarshkshah1992) [inline caboose.go:156]: So this API will ONLY be used for fetching CARs, correct ? Can I give a path such as just /ipfs/{cid} to it ? 

Will that give me the entire CAR rooted at that cid or just the block  for that cid ? The Saturn URL in that case will be {Saturn Node IP}/ipfs/cid.  Does Saturn have an API in place already to be able to tell that it should return the entire CAR for that root and not just the block ? Is it the `format=raw?` part in `saturnReqTmpl` that indicates that we want to fetch just a block ?

File: caboose.go
Comment (aarshkshah1992) [inline caboose.go:155]: Please can you add some e2e tests for `Fetch` just like we have for `Get` ?

File: pool.go
Comment (aarshkshah1992) [inline pool.go:6]: Please do rebase as many changes to pool.go have been merged to main already.

File: pool.go
Comment (aarshkshah1992) [inline pool.go:376]: Where do we put more stuff in `pq` to be able to do this here ?

File: pool.go
Comment (aarshkshah1992) [inline pool.go:392]: Can combine `fetchResourceAndUpdate` and `fetchBlockAndUpdate` into one function by making "doFetch"/"fetchResource" an argument to it.

File: pool.go
Comment (aarshkshah1992) [inline pool.go:366]: When would a callback return `ErrPartialResponse ` ? Is it when I download a CAR -> get some blocks and am able to verify them -> but then validation fails on other blocks or I am simply not able to get them ? In that case, where are the additional `paths` in `pq` coming from here ? Could you give an example of what this additional `StillNeed` paths in `pq` would look like ?

File: pool.go
Comment (aarshkshah1992) [reply pool.go:376]: I think the `pq = append(pq, epe.StillNeed...)` in the line below should happen before line 376 ?

File: pool.go
Comment (aarshkshah1992) [inline pool.go:340]: could panic. We should check the length first ?

**Review by aarshkshah1992 — COMMENTED** [review]
First round of review. Lots of questions as I don't understand how Bifrost is gonna use this API. Some examples/tests for this new Fetch API/how the callback would work if everything goes right/how the call -back would leverage the `ErrPartialResponse`/what would happen if the callback isn't able to incrementally verify what's coming in from the response would be great.

**Comment by willscott** [comment]
can you confirm my merge looks reasonable?

File: fetcher.go
Comment (willscott) [reply fetcher.go:67]: this was previously present for blocks  - this is not new code but has been separated out from pool https://github.com/filecoin-saturn/caboose/blob/main/pool.go#L544

File: fetcher.go
Comment (willscott) [reply fetcher.go:66]: a car, yeah `/ipfs/cid/path/to/file/or/dir` with accept of car will return a car

File: fetcher.go
Comment (willscott) [reply fetcher.go:207]: yes, it's responsible for verifying. if it gets a partial failure it errors out at that point with the new error.

File: caboose.go
Comment (willscott) [reply caboose.go:226]: > So this API will ONLY be used for fetching CARs, correct ?
yes
> Can I give a path such as just /ipfs/{cid} to it ?
yes
> Will that give me the entire CAR rooted at that cid or just the block for that cid ? 
that's up to the backing saturn node to implement the semantics defined at
https://www.notion.so/pl-strflt/Graph-Request-APIs-97cc90e74df44404a0f9aa29d7291a5f

File: pool.go
Comment (willscott) [reply pool.go:492]: it doesn't matter - this one is removing the current request that was just fetched

File: pool.go
Comment (willscott) [reply pool.go:456]: pq[0] is referenced on line 337, there is no case at the end of the loop where pq will have length 0, right?

File: metrics.go
Comment (lidel) [inline metrics.go:71]: :thought_balloon:  we will need separate histogram buckets for CAR requests. Fetching bigger dags will take longer than 19s or 60s, may  run for arbitrary lengths of time.

I don't know what is the sensible max, we may start with 30min and look at the distribution once the data starts flowing:

```go
durationPerCarHistogram = prometheus.ExponentialBucketsRange(50, 1800000, 10)
```

Given the ambiguous range, I feel we can reuse the same bucket definition for both `duration_car_success` and `duration_car_peer_success`

**Review by lidel — CHANGES_REQUESTED** [review]

File: caboose.go
Comment (aarshkshah1992) [inline caboose.go:228]: Please can we have a e2e test for `Fetch` ?

File: pool.go
Comment (aarshkshah1992) [inline pool.go:571]: Please can you add a test for `updateWeightBatched`

Comment (willscott) [reply pool.go:571]: you added (and maybe later removed) this
https://github.com/filecoin-saturn/caboose/blob/67d62dd0ca18af3137522a6ae103b538f2923cb3/pool.go#L329

File: pool.go
Comment (aarshkshah1992) [inline pool.go:437]: Please replace the above code with `getNodesToFetch`. You can take a look at how we do it in `fetchBlockWith`

File: pool.go
Comment (aarshkshah1992) [reply pool.go:571]: Okay I'll add the test back.

**Comment by aarshkshah1992** [comment]
@willscott Please just address the two remaining items and we can merge !

**Comment by aarshkshah1992** [comment]
@willscott Please can you address the unresolved items before we merge this ?

**Comment by willscott** [comment]
i wonder if you can take a pass through the weighting thing and get it somewhere you're happy with?

**Comment by aarshkshah1992** [comment]
@willscott Wonder not for I can and I will.

**Comment by aarshkshah1992** [comment]
Please hold off the merge till I finish AB testing.

**Comment by aarshkshah1992** [comment]
```
aarshshah@Aarshs-MacBook-Pro-2 caboose % ab -l -k -n 10000 -c 1000 -w "http://localhost:8081/ipns/en.wikipedia-on-ipfs.org/wiki/" 
<p>
 This is ApacheBench, Version 2.3 <i>&lt;$Revision: 1879490 $&gt;</i><br>
 Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/<br>
 Licensed to The Apache Software Foundation, http://www.apache.org/<br>
</p>
<p>
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


<table >
<tr ><th colspan=2 bgcolor=white>Server Software:</th><td colspan=2 bgcolor=white></td></tr>
<tr ><th colspan=2 bgcolor=white>Server Hostname:</th><td colspan=2 bgcolor=white>localhost</td></tr>
<tr ><th colspan=2 bgcolor=white>Server Port:</th><td colspan=2 bgcolor=white>8081</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Path:</th><td colspan=2 bgcolor=white>/ipns/en.wikipedia-on-ipfs.org/wiki/</td></tr>
<tr ><th colspan=2 bgcolor=white>Document Length:</th><td colspan=2 bgcolor=white>Variable</td></tr>
<tr ><th colspan=2 bgcolor=white>Concurrency Level:</th><td colspan=2 bgcolor=white>1000</td></tr>
<tr ><th colspan=2 bgcolor=white>Time taken for tests:</th><td colspan=2 bgcolor=white>30.057 seconds</td></tr>
<tr ><th colspan=2 bgcolor=white>Complete requests:</th><td colspan=2 bgcolor=white>10000</td></tr>
<tr ><th colspan=2 bgcolor=white>Failed requests:</th><td colspan=2 bgcolor=white>0</td></tr>
<tr ><th colspan=2 bgcolor=white>Non-2xx responses:</th><td colspan=2 bgcolor=white>10000</td></tr>
<tr ><th colspan=2 bgcolor=white>Keep-Alive requests:</th><td colspan=2 bgcolor=white>10000</td></tr>
<tr ><th colspan=2 bgcolor=white>Total transferred:</th><td colspan=2 bgcolor=white>7188641 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>HTML transferred:</th><td colspan=2 bgcolor=white>3230416 bytes</td></tr>
<tr ><th colspan=2 bgcolor=white>Requests per second:</th><td colspan=2 bgcolor=white>332.70</td></tr>
<tr ><th colspan=2 bgcolor=white>Transfer rate:</th><td colspan=2 bgcolor=white>233.56 kb/s received</td></tr>
<tr ><th bgcolor=white colspan=4>Connection Times (ms)</th></tr>
<tr ><th bgcolor=white>&nbsp;</th> <th bgcolor=white>min</th>   <th bgcolor=white>avg</th>   <th bgcolor=white>max</th></tr>
<tr ><th bgcolor=white>Connect:</th><td bgcolor=white>    0</td><td bgcolor=white>    2</td><td bgcolor=white>   40</td></tr>
<tr ><th bgcolor=white>Processing:</th><td bgcolor=white>    2</td><td bgcolor=white> 1970</td><td bgcolor=white>21292</td></tr>
<tr ><th bgcolor=white>Total:</th><td bgcolor=white>    2</td><td bgcolor=white> 1972</td><td bgcolor=white>21332</td></tr>
</table>

```

**Pull request merged**
